sdk/core/core-https/src/nodeHttpsClient.ts

```diff
@@ -0,0 +1,293 @@
+  /**
+   * Makes a request over an underlying transport layer and returns the response.
+   * @param request The request to be made.
+   */
+  public async sendRequest(request: PipelineRequest): Promise<PipelineResponse> {
+    const abortController = new AbortController();
+    let abortListener: ((event: any) => void) | undefined;
+    if (request.abortSignal) {
+      if (request.abortSignal.aborted) {
+        throw new AbortError("The operation was aborted.");
+      }
+
+      abortListener = (event: Event) => {
+        if (event.type === "abort") {
+          abortController.abort();
+        }
+      };
+      request.abortSignal.addEventListener("abort", abortListener);
+    }
+
+    if (request.timeout > 0) {
+      setTimeout(() => {
+        abortController.abort();
+      }, request.timeout);
+    }
+
+    if (request.formData) {
+      prepareFormData(request.formData, request);
+    }
+
+    if (!request.skipDecompressResponse) {
+      request.headers.set("Accept-Encoding", "gzip,deflate");
+    }
+
+    let body = request.body;
+
+    if (body && request.onUploadProgress) {
+      const onUploadProgress = request.onUploadProgress;
+      const uploadReportStream = new ReportTransform(onUploadProgress);
+      if (isReadableStream(body)) {
+        body.pipe(uploadReportStream);
+      } else {
+        uploadReportStream.end(body);
+      }
+
+      body = uploadReportStream;
+    }
+
+    if (body) {
+      if (isReadableStream(body)) {
+        request.headers.set("Transfer-Encoding", "chunked");
+      } else if (typeof body === "string") {
+        request.headers.set("Content-Length", body.length);
+      }
+    }
+
+    try {
+      const result = await new Promise<PipelineResponse>((resolve, reject) => {
+        const options = getRequestOptions(request);
+        const req = https.request(options, async (res) => {
+          const headers = getResponseHeaders(res);
+
+          const status = res.statusCode ?? 0;
+          const response: PipelineResponse = {
+            status,
+            headers,
+            request
+          };
+
+          let responseStream = getResponseStream(res, headers, request.skipDecompressResponse);
+
+          const onDownloadProgress = request.onDownloadProgress;
+          if (onDownloadProgress) {
+            const downloadReportStream = new ReportTransform(onDownloadProgress);
+            responseStream.pipe(downloadReportStream);
+            responseStream = downloadReportStream;
+          }
+
+          if (request.streamResponseBody) {
+            response.readableStreamBody = responseStream;
+          } else {
+            response.bodyAsText = await streamToText(responseStream);
+          }
+
+          resolve(response);
+        });
+        req.on("error", (err) => {
+          reject(new RestError(err.message, { code: RestError.REQUEST_SEND_ERROR, request }));
+        });
+        abortController.signal.addEventListener("abort", () => {
```

Comment:
We also need to cleanup this event listener or else we may end up with a memory leak if the user uses the same signal in multiple operations.

Reply:
This isn't the signal that was passed in on the request though, it's a controller we create inside the method. We do clean up unregistering from request.abortSignal later

Reply:
Ah shoot, sorry you're right, I missed that this was on the abortController we create.